### ISSUE TITLE
Proposal: inspect image json on a registry

### ISSUE THREAD

**Issue by runcom** [docker/distribution#1252]
TL;DR;
---------
Add the ability to pull down an image JSON from a registry w/o the need to download all layers.
This is trying to carry docker/docker#14257 and docker/docker#14258. The latter was closed since `docker inspect` should not interact with registries.

`docker/distribution/cmd/inspect`
-----------
What I'm thinking should be pretty simple to implement (I hope) but I decided to write down this proposal to gather feedback instead of going straight to the code.

I agree adding the notion of *remote* to `docker inspect` is not doable (since just a small subset of docker commands deal with *remotes* and docker doesn't want to couple other commands to remotes). Adding an option to *shallow* pull an image could be good but that's definitely much work for something, IMO, easy like pulling a manifest and format it.

The main use case we are looking for with this is the ability to verify an image is up date.
The common use case is to be able to download and examine one or more image JSON files from the registry to discover which images you want to pull.
Our goal it to look at the `LABEL` data associated with images. Examples may include, show me all of the images based on rhel7, show me all of the images with this license field and so on. I'm sure others may have other use cases I failed to list here now.

In the docker issues mentioned above, a proposal was to have an external tool that can inspect a remote registry.
I was thinking if an external command could be added under `docker/distribution/cmd/`.
I'd call it `inspect` and its work should be just to inspect images on registries returning the JSON associated with the given image.

```
$ inspect docker.io/fedora
[
{
    "Id": "ded7cd95e059788f2586a51c275a4f151653779d6a7f4dad77c2bd34601d94e4",
    "Parent": "48ecf305d2cf7046c1f5f8fcbcd4994403173441d4a7f125b1bb0ceead9de731",
    ...
    "Os": "linux",
    "Size": 186515831
}
]

$ inspect ubuntu
LOG: using official docker.io registry
[
{
    "Id": "ded7cd95e059788f2586a51c275a4f151653779d6a7f4dad77c2bd34601d94e4",
    "Parent": "48ecf305d2cf7046c1f5f8fcbcd4994403173441d4a7f125b1bb0ceead9de731",
    ...
    "Os": "linux",
    "Size": 186515831
}
]

[...]
```

@stevvooe I noticed you were pretty active on having something like this, I'd love to hear something from you
@aaronlehmann @tonistiigi what do you guys think?

**Comment by aaronlehmann** [comment]
I like the idea of having such a tool and think it would be straightforward to implement.

The manifest refactor work that @RichardScothern is doing in https://github.com/docker/distribution/pull/1105 would help do this in a forward-compatible way. The inspect tool would first call `Target` on a manifest, to see if there's a single image configuration JSON referenced from the manifest (as is the case with the new manifest format which is being worked on; see #1068). If `Target` returns nil, it could attempt a type assertion to `*schema1.SignedManifest` (the current manifest format), and if that succeeds, use the JSON from the first `V1Compatibility` field (@RichardScothern: is this the right approach?).

**Comment by runcom** [comment]
@aaronlehmann do you see #https://github.com/docker/distribution/pull/1068 as a blocker for a first implementation of such tool?

**Comment by aaronlehmann** [comment]
Not necessarily a blocker, but the tool would need some modifications after #1068.

**Comment by tonistiigi** [comment]
Not 100% sure why this can't be an external tool. For example, I think the hub default location shown in your example is docker specific and not something that distribution knows about. If you want to make sure the image is up-to-date you should compare the image IDs, comparing outputs of `distribution inspect` and `docker inspect` is not something that you can rely on.

> Examples may include, show me all of the images based on rhel7, show me all of the images with this license field and so on. I'm sure others may have other use cases I failed to list here now.

How can this be implemented with a simple inspect based on a single image?

**Comment by runcom** [comment]
> I think the hub default location shown in your example is docker specific and not something that distribution knows about.

that was just because I wanted to keep the default namespace as it's in docker..I think ppl are so used to that..

> How can this be implemented with a simple inspect based on a single image?

We're relying on `LABEL` a lot and we store all such things there

> comparing outputs of distribution inspect and docker inspect is not something that you can rely on.

We will just inspect labels on disk and labels remote

-

I would to put it under cmd here because it sounds specific to distribution as discussed in the linked docker issues (sounds because it deals with registries) and it will be just `inspect` so it's just using APIs here

**Comment by rhatdan** [comment]
We have added a tool to the atomic command called verify.

atomic verify mongodb-image

Which would examine the labels in mongodb-image and all of the images it was based on for labels.

Imagine 
mongodb-image
Had a label
LABEL NAME=mongodb-image
LABEL VERSION=1.2
LABEL RELEASE=1
And it was based on rhel7 base image with
LABEL NAME=rhel7
LABEL VERSION=7.1
LABEL RELEASE=1

We want the tool to first search for a new version of mongodb-image on a remote registery, if the remote registry had
LABEL NAME=mongodb-image
LABEL VERSION=1.3
LABEL RELEASE=1
Then atomic verify would report their is a new version available, but more interesting would be looking at the sub images and reporting back that mongodb-image is latest but their is a newer base image for RHEL7
LABEL NAME=rhel7
LABEL VERSION=7.2
LABEL RELEASE=1

We recommend that you contact mongodb-image provider to rebuild its image, or if you built it you should rebuild it using docker build.

**Comment by stevvooe** [comment]
@runcom The main blocker to this is the amount of work we are doing in distribution. The [moratorium](https://github.com/docker/docker/blob/master/ROADMAP.md#23-remote-registry-operations) is related to this and applies somewhat to the distribution project, at least thematically. Supporting this in a backwards compatible manner will slow down advancement in the image manifest. We'd also like to minimize the exposure of the current manifest format, if possible.

I would recommend stepping back and looking harder at the actual schema with a proposal like this. The examples you have are showing image configuration json. The distribution project doesn't really crack open this format, nor should it understand the details of the distributed content.

There are two use cases cited in this proposal that don't really lend itself to this solution. First is that polling whether something is up to date is really inefficient. The second is that labels don't really matter at the distribution level. There is a notification system that can support both of these without incurring extra request load on images or any code in this project. One simply has to listen to manifest notifications, fetch the content and index by label or however works for you. I have seen several proposals in this vein from redhat and haven't seen an analysis of why this approach will not work.

All in all, the goal of the docker distribution project has been to decouple image format from distribution. I am not sure if such a command is compatible with those goals. This goal needs to be addressed directly in anything that further couples the layer model with the registry.

Any successful proposal in this area would have the following properties:

1. Primary object is a manifest.
1. Remote registry object inspection: download manifests, layers individually, by reference.
2. Tools to unpack and interpret content (imagine `registry inspect ubuntu:latest --format {{range .Layers}}.ID\n{{end}}`). Would have to provide a "common" template structure to work across manifest versions.
3. Allow one to specify the targeted registry.
4. De-coupling with image format and registry must be strictly maintained.

Optionally:

1. Structure this as a subcommand of `registry`.
2. Better auth management than docker (not a docker.cfg file in plain text in the home directory).

**Comment by rhatdan** [comment]
Where does one listen to these notifications, should every machine in the field listen for notifications?  As an admin I want to verify that all of my images are up2date, have no security vulnerabilities, and have not been mucked with.  These are the goals we are after.  How does my security officer run a check to make sure this is the case?

**Comment by runcom** [comment]
> Tools to unpack and interpret content (imagine `registry inspect ubuntu:latest --format {{range .Layers}}.ID\n{{end}})`. Would have to provide a "common" template structure to work across manifest versions.

would `registry inspect ubuntu:latest --format '{{ Labels }}'` be ok maintaining backward compatibility between manifest versions with a common template? (just to be sure I understand)

**Comment by stevvooe** [comment]
@runcom Yes.

@rhatdan This is the third time we've had a similar proposal from your team members. Rather than closing it out, we are considering the possibility. I'd at least like to see a study of the alternatives before adding more code that we are going to have to maintain into the future.

Something that involves notary is probably much more in line with the goals you have a above. Rather than dismissing this possibility out of hand, please explore the available features with an open mind.

**Comment by rhatdan** [comment]
I am not dismissing what you are saying, I just don't see how it matches.  I feel like we are talking about apples and oranges.  I am not trying to confirm that the image came from a particular company, or are signed or are trusted.

I am just trying to see if there is a newer image available.  We put the name/version/release into the image json data as LABELS.  All we want is access to this json data at the Registries.  Might be best to do this in Bluejeans or via IRC.

Using something like Notary to make sure we trust the LABELS might be useful in the future.

**Comment by stevvooe** [comment]
> I am just trying to see if there is a newer image available.

Polling gives you absolutely no guarantee of freshness or consistency.

"Freshness Guarantees" is the second item in https://github.com/docker/notary#goals. Notary gives the most recent set of images that are signed. They can provide roots to get access to the manifests, which are more than accessible via the API.

Notifications can also provide decent guarantees of freshness, albeit any implementation would suffer from some of the attacks that TUF handles. If you have no intention of using notary right now, I would suggest you explore this option.

**Comment by rhatdan** [comment]
You keep calling this polling, indicating that you think we are having a constant polling of the registry about new versions available.  I see this more like a scan, directed by a human to look for updates.  I would like to take advantage of new features of Notary if they could help out, but Notary will not be used everywhere.  Customers will have their own registries.

How do I get these `Notifications`?  If I miss a `Notification` or am not listening, how can I tell that my app is based on an old version of a base image?

**Comment by runcom** [comment]
I assume the notifications come from the registry itself

**Comment by mtrmac** [comment]
## What Notary Does

For the use case in https://github.com/docker/distribution/issues/1252#issuecomment-163357529 , the notary can efficiently and securely¹ determine that $id_A is the currently latest version of `mongodb-image`; this could be periodically verified by a scan/audit process, or perhaps (https://github.com/docker/docker/issues/16654) allow enforcing a policy of never pulling obsolete images.

The notary can similarly determine that $id_B is *not* the currently latest version of `rhel7`.

The notary can *not* determine that $id_A is dependent on $id_B, and that it should be rebuilt on top of an updated `rhel7` image; the notary does not inspect the contents of the signed manifests at all and has no knowledge of Docker image format and semantics.

## Use Case

Per my (limited) understanding, an audit/scan mechanism would:

0. (Get a list of “interesting” tags from $somewhere (out of scope for this issue), and for each one:)
1. Ask the notary for the latest metadata, providing a signed mapping from tags to image IDs. (That would be `mongodb-image`→$id_A in the example above.)
2. (Verify that the relevant machines are running containers only based on image with $id_A and not some older one. Out of scope for this issue.)
3. (Possibly do some other kind of scan/audit on the image $id_A which I have no idea about and is out of scope for this issue.)
4. Using the content-addressed $id_A, ask the registry for the JSON manifest.
5. From the JSON, using some mechanism (which I treat as unspecified and out of scope for this issue, but @rhatdan and others probably know what), figure out the $base_tag and $base_ID on which $id_A is based (that would be `rhel7` and $id_B in the example above).
6. Ask the notary for the latest metadata of $base_tag and verify that $base_ID is the latest version of $base_tag.
7. Perhaps take $base_ID and start again from step 4. above, following the “base image" chain all the way to the lowest layer.

# Implementation Placement

Step 5 (and 6?) is what this issue is about; there needs to be a registry client which can fetch the image manifest. Others like @rhatdan can probably speak better to whether this should be a dumb fetch tool or something which directly understands the Docker JSON format, and to what extent; I know little, all I can say that it needs to fetch only the JSON and not the gigabytes of tar files.

If, as you seem to argue, steps 5+6 do not belong into the registry or distribution, they definitely do not belong into the notary either: notary only cares about lengths and SHA256 hashes of the manifests, it does not store them at all (so it cannot parse them even if it wanted). (Also, the notary is already signing the $id_A mapping, so the JSON used in steps 5+6 is sufficiently authenticated by notary. It does not need to be involved beyond that.)

---

(Non-essential comments, feel free to skip:)

## Polling

As for having to poll, notary does not help there noticeably; AFAICS it does not have any kind of event notification or message bus facility, the audit mechanism would still have to periodically poll the notary for changes. The only difference between the notary and registry in this respect is that the notary design allows just downloading two small files per repository to verify that nothing has changed since last time, instead of fetching the JSON for each tag in the repository from the registry individually; but that is only useful if the audit mechanism is keeping state across runs, and it is a smallish optimization in any case.

## Freshness

The freshness discussion seems to have kind of derailed. At the moment¹ both the notary and just talking to the registry over HTTPS gives a comparable cryptographic guarantee that the answer is relatively recent (the major difference being how much you trust the two servers, their storage backends, and key storage mechanisms.) Going beyond “relatively recent” gets unrealistic very fast: A client can at T+0 ask the {registry,notary} for the latest tag mapping, at T+1 the registry will send an answer saying “version 1”, at T+2 the repository author uploads a new image version 2, and at T+3 the client will receive a signed answer that version 1 is the latest version when that is, strictly speaking, no longer the case.

## Consistency

And finally, the TUF consistency guarantee is mostly a red herring in the context of Docker: It guarantees that I see “consistent” mapping of tags to IDs for `rhel7:latest`, `rhel7:7.0`, `rhel7:7.1` and`rhel7:7.1.1`, which does not have any real-world meaning at all. (It could, *perhaps*, have a real-world meaning for a consistent snapshot of `full-stack-app:database`, `full-stack-app:business-layer` and `full-stack-app:web-frontend`, but this kind of consistency is not really a security concern, and is anyone using this kind of naming? And anyway, when migrating such a tightly coupled stack, image naming is not really a pressing concern AFAICT; once the hard problem of migration and uptime is solved, it can be executed with any names or just using raw image IDs.)

---

### Off-topic

¹ Well, ”securely”… given the current TOFU+trust-every-author design, notary is not currently *noticeably* more secure than just asking the registry over HTTPS. But when pull-through registry mirrors start to appear, and notary gains some trust configuration policy mechanism, notary will be superior.

**Comment by stevvooe** [comment]
@mtrmac Thanks for the analysis.

> The notary can not determine that $id_A is dependent on $id_B, and that it should be rebuilt on top of an updated rhel7 image; the notary does not inspect the contents of the signed manifests at all and has no knowledge of Docker image format and semantics.

If I implied this above, that was unintentional. The requirement was to have the current set of manifests for a given repository. Notary can assist in this. Further processing would be required to fetch label-based dependency information over the API.

Notary _does_ provide a single-point transaction barrier that may be useful. This may be desirable in the face of changes that may occur in the registry.

> If, as you seem to argue, steps 5+6 do not belong into the registry or distribution, they definitely do not belong into the notary either: notary only cares about lengths and SHA256 hashes of the manifests, it does not store them at all (so it cannot parse them even if it wanted).

I'm not sure where I said this, if this is directed at me. Please go back and carefully read what I said, as I think you are misunderstanding several points.

@runcom 
> I thought you agreed on a proposal for a common template and a registry inspect command

I provided an outline for a possibly successful proposal. I do like the notion of an inspect command, for registry objects, but I am worried that it is not at all the right solution for the intended use case. It would be much more efficient to have a notification listener index manifests by their labels and perform dependency analysis in response to messages.

@rhatdan We have found solid reliability with the notification stack. We've implemented a lot of functionality based on it with success. We did discuss the issue of adding more reliable notification sinks, but haven't had much need. Modifying it to add support for a more reliable backend would be extremely straightforward, if your requirements are different.

**Comment by qianzhangxa** [comment]
Hi Guys,

Is there any work ongoing or progress on this proposal? I think our use case will also benefit from this proposal: we'd like Docker registry provide a REST API with which we can get the JSON description for a specific image, and check the "architecture" field to see if it is a amd64 image or ppc64le image or others, and then do our own logics based on this information.

**Comment by RichardScothern** [comment]
@qianzhangxa I think your use case is covered with the current registry API.  While there is no tool, you can fetch the runtime configuration from the registry with the current API.  

The v2 manifest will show the digest of the runtime configuration, which when fetched will will reveal the architecture.  Have a look at https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md

**Comment by runcom** [comment]
@qianzhangxa you can use https://github.com/projectatomic/skopeo which ease what you need

**Comment by stevvooe** [comment]
@qianzhangxa Would have loved to have a contribution in this area.

https://github.com/docker/docker/issues/24739 is a great proposal in this area.

**Comment by qianzhangxa** [comment]
Thanks guys! @RichardScothern I have a Docker registry v2 running, can you please let me know which specific API I can use to get manifest of a Docker image?

**Comment by RichardScothern** [comment]
start here https://docs.docker.com/registry/spec/api/#/get-manifest @qianzhangxa

**Comment by qianzhangxa** [comment]
That works perfectly! Thanks @RichardScothern !

**Comment by jhmartin** [comment]
There are other reasons to fetch image metadata from a repository -- a deployment tool might want to see something about the image; maybe the metadata defines a desired memory reservation.  Fetching the entire image to gather this data is needlessly heavy.

**Comment by kaos** [comment]
I'm looking for a way to do exactly what @jhmartin is talking about. In my case, I want to find out the ENTRYPOINT for an image, without downloading any layers.

**Comment by stevvooe** [comment]
@kaos @jhmartin The registry already supports fetching manifest and configuration. What exactly are you looking for?

**Comment by kaos** [comment]
I'm not sure what the configuration holds, but the manifest doesn't give the same info as what `docker inspect` does. Not that I need _all_ that information, and for the ENTRYPOINT, I _could_ parse the RUN command for each layer, but that just feels clumsy.

@stevvooe 
> What exactly are you looking for?

Wasn't `I want to find out the ENTRYPOINT for an image, without downloading any layers.` exact enough?

Thanks.

**Comment by dmcgowan** [comment]
@kaos this is already possible doing exactly what Stephen suggested, download the manifest and configuration, the entrypoint will be in the configuration. That is exactly what a tool would end up doing.

**Comment by kaos** [comment]
@dmcgowan Ok, that's awesome. But where do I get the configuration part?
I'm reading https://docs.docker.com/registry/spec/api/ and see `GET /v2/<name>/manifests/<reference>` for reading the image manifest, but haven't found anything regarding any configuration. Am I looking in the wrong place?

**Comment by stevvooe** [comment]
@kaos Everything is there. That is why I ask "What exactly are you looking for?".

The configuration is just a blob pointed to by the manifest. Just fetch the configuration with the digest against `GET /v2/<name>/blobs/<reference>`.

**Comment by dmcgowan** [comment]
The API flow to get an entrypoint from the hub would look something like....
```
IMAGE=library/redis
TAG=latest
TOKEN=$(curl -s "https://auth.docker.io/token?scope=repository:$IMAGE:pull&service=registry.docker.io" | jq -r .token)
CONFIG_DIGEST=$(curl -s -H"Accept: application/vnd.docker.distribution.manifest.v2+json" -H"Authorization: Bearer $TOKEN" "https://registry-1.docker.io/v2/$IMAGE/manifests/$TAG" | jq -r .config.digest)

ENTRYPOINT=$(curl -sL -H"Authorization: Bearer $TOKEN" "https://registry-1.docker.io/v2/$IMAGE/blobs/$CONFIG_DIGEST" | jq -r .container_config.Entrypoint)

echo $ENTRYPOINT
```

**Comment by kaos** [comment]
Thanks, yes I found it now. Being a bit dumb (new to the registry api), I failed to pass the accept header, so I got a v1 manifest, which didn't have the configuration part. The [docs](https://docs.docker.com/registry/spec/manifest-v2-2/) were't very clear about the configuration part either (had to read it really carefully..).

This was the piece of information nudging me in the right direction:
> The configuration is just a blob pointed to by the manifest. Just fetch the configuration with the digest against GET /v2/<name>/blobs/<reference>.

So, all good. Thank you for your patience :)

**Comment by kaos** [comment]
@dmcgowan wow, very instructive reply 👍 .  That would've worked perfectly too (didn't see it until I came back to type my findings, though..)

**Comment by jhmartin** [comment]
@stevvooe does the same work for fetching image metadata / [LABEL](https://docs.docker.com/engine/reference/builder/#/label)'s of an image?

**Comment by stevvooe** [comment]
@jhmartin Anything in the regular image config will be there, which includes labels.

**Comment by jhmartin** [comment]
Awesome!

**Comment by dailyherold** [comment]
This thread was very helpful, thank you everyone!
